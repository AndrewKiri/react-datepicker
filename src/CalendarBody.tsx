import * as React from 'react';
import * as moment from 'moment';
import { getDayMatrix, isDayEqual } from './util/DateUtil';
import Day from './Day';

interface Props{
  current: moment.Moment
  selected?: moment.Moment[]
  startDay?: moment.Moment
  endDay?: moment.Moment
  onChange?: (date: moment.Moment) => void
  customDayClass?: (date: moment.Moment) => string | string[]
  customDayText?: (date: moment.Moment) => string
}
class CalendarBody extends React.Component<Props> {
  
  constructor(props:Props){
    super(props);
  }
  getCustomClass = (date: string):string => {
    const { current, customDayClass } = this.props;
    const currentDate = moment(current).date(parseInt(date));
    let classArr:string[] = [];
    
    if(!date.trim()) return '';

    classArr.push(`calendar__day--${currentDate.day()}`);

    if( customDayClass !== undefined ) {
      const customClass = customDayClass(currentDate);
      classArr = classArr.concat(typeof(customClass) === 'string' ? 
        [customClass]: customClass);
    }

    return classArr.join(' '); 
  }

  getCustomText = (date: string):string => {
    const { current, customDayText, startDay, endDay} = this.props;
    const currentDate = moment(current).date(parseInt(date));
    if(!date.trim()) return '';
    if(!customDayText) return '';

    return customDayText(currentDate);
  }

  isSelected = (date: string):boolean => {
    const { selected } = this.props;
    if(selected === undefined) return false;
    return selected.some(v => this.isDayEqual(date, v));
  }

  isDayEqual = (date: string, compareDate?: moment.Moment):boolean => {
    const { current } = this.props;
    const currentDate = moment(current).date(parseInt(date));
    if(!date.trim()) return false;
    if(!compareDate) return false;
    return isDayEqual(compareDate, currentDate);
  } 

  isRange = (date: string):boolean => {
    const { current, startDay, endDay } = this.props;
    const currentDate = moment(current).date(parseInt(date));
    
    if(!date.trim()) return false;
    if((startDay && currentDate.diff(startDay, 'days') > 0 )
      && (endDay && currentDate.diff(endDay, 'days') < 0)) return true;

    return false;
  }


  handleChange = (value: string) => {
    const { onChange, current } = this.props;

    if(onChange)
      onChange(moment(current).date(parseInt(value)));
  }

  render() {
    const {
      current,
      startDay,
      endDay
    } = this.props;

    const dayMatrix = getDayMatrix(current.year(), current.month());
    const weekdays = moment.weekdaysShort();

    return (
      <div className="calendar__body">
        <table className="calendar__body--table">
          <thead>
            <tr>
              {
                weekdays.map((v,i) => <th key={i}>{v}</th>)
              }
            </tr>
          </thead> 
          <tbody>
            {
              dayMatrix.map((row,i) => (
              <tr key={i}>
                  {
                    row.map((date,j) => (
                      <Day 
                        customClass={this.getCustomClass(date)}
                        customText={this.getCustomText(date)}
                        onChange={this.handleChange}
                        start={this.isDayEqual(date, startDay)}
                        end={this.isDayEqual(date, endDay)}
                        range={this.isRange(date)}
                        today={this.isDayEqual(date, moment())}
                        selected={this.isSelected(date)}
                        value={date} 
                        key={i+j}
                      />
                    ))     
                  }
                </tr>
              ))
            }
          </tbody>
        </table>
      </div>
  )
  }
}
export default CalendarBody;