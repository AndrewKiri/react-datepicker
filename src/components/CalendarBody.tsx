import * as moment from 'moment';
import * as React from 'react';
import { CalendarEnums } from '../common/@enum';
import { DatePickerDefaults } from '../common/Constant';
import { getMonthMatrix, getYearMatrix } from '../utils/DateUtil';
import DayView, { Props as DayViewProps } from './DayView';
import TableCell from './TableCell';
import TableMatrixView from './TableMatrixView';

interface CalendarBodyProps {
  /** Calendar viewMode(Year, Month, Day) */
  viewMode: CalendarEnums.ViewMode;
  /** Calendar current Date */
  current: moment.Moment;
  /** DayClick Event */
  onClick: (value: string) => void;
  /** Locale to use */
  locale: string;
}
type Props = DayViewProps & CalendarBodyProps;
class CalendarBody extends React.Component<Props> {
  public static defaultProps = {
    viewMode: CalendarEnums.ViewMode.DAY,
    onClick: () => {},
    locale: DatePickerDefaults.locale,
  };

  public render() {
    const { current, onClick, locale } = this.props;
    const cell = (className: string) => (value: string, key: number) => (
      <TableCell className={className} text={value} onClick={text => onClick(text)} key={key} />
    );
    const viewMap = {
      [CalendarEnums.ViewMode.YEAR]: (
        <TableMatrixView matrix={getYearMatrix(current.year())} cell={cell('calendar__year')} />
      ),
      [CalendarEnums.ViewMode.MONTH]: (
        <TableMatrixView matrix={getMonthMatrix(locale)} cell={cell('calendar__month')} />
      ),
      [CalendarEnums.ViewMode.DAY]: <DayView {...this.props} />,
    };

    return <div className="calendar__body">{viewMap[this.props.viewMode]}</div>;
  }
}
export default CalendarBody;
