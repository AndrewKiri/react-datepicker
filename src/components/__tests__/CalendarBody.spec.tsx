import { mount, shallow } from 'enzyme';
import * as moment from 'moment';
import * as React from 'react';
import * as sinon from 'sinon';
import CalendarBody from '../CalendarBody';

const mockMoment = moment.unix(1543622400);

describe('<CalendarBody />', () => {
  it('props current correctly', () => {
    const component = shallow(<CalendarBody current={mockMoment} />);

    expect(component).toMatchSnapshot();
  });

  it('props selected correctly', () => {
    const selected = [
      moment('20181202'),
      moment('20181205'),
      moment('20181212'),
      moment('20181222'),
    ];
    const component = mount(<CalendarBody current={mockMoment} selected={selected} />);

    expect(component).toMatchSnapshot();
    expect(component.find('td.calendar__day--selected')).toHaveLength(selected.length);
  });

  it('props startDay, endDay correctly', () => {
    const startDay = moment('20181205', 'YYYYMMDD');
    const endDay = moment('20181212', 'YYYYMMDD');

    const component = mount(
      <CalendarBody current={mockMoment} startDay={startDay} endDay={endDay} />
    );

    expect(component).toMatchSnapshot();
    expect(component.find('td.calendar__day--start')).toHaveLength(1);
    expect(component.find('td.calendar__day--end')).toHaveLength(1);
    expect(component.find('td.calendar__day--start > div').text()).toEqual('5');
    expect(component.find('td.calendar__day--end > div').text()).toEqual('12');
    expect(component.find('td.calendar__day--range')).toHaveLength(6);
  });

  it('props onClick correctly', () => {
    const onClick = sinon.spy();

    const component = mount(<CalendarBody current={mockMoment} onClick={onClick} />);

    expect(component).toMatchSnapshot();
    component
      .find('td')
      .first()
      .simulate('click');
    expect(onClick).toHaveProperty('callCount', 1);
  });

  it('props customDayClass correctly', () => {
    const customDayClass = (date: moment.Moment) => {
      // custom day class string or array
      const dayClassMap = {
        '20181202': ['custom-day', 'day-test1', 'day-test2'],
        '20181211': 'custom-day',
      };
      return dayClassMap[date.format('YYYYMMDD')];
    };

    const component = mount(<CalendarBody current={mockMoment} customDayClass={customDayClass} />);

    expect(component).toMatchSnapshot();
    expect(component.find('td.day-test1')).toHaveLength(1);
    expect(component.find('td.day-test2')).toHaveLength(1);
    expect(component.find('td.day-test1').text()).toEqual('2');
    expect(component.find('td.custom-day')).toHaveLength(2);
  });

  it('props customDayText correctly', () => {
    const customDayText = (date: moment.Moment) => {
      // custom day class string or array
      const dayTextMap = {
        '20181202': '신정',
        '20181211': '공휴일',
      };
      return dayTextMap[date.format('YYYYMMDD')];
    };

    const component = mount(<CalendarBody current={mockMoment} customDayText={customDayText} />);

    expect(component).toMatchSnapshot();
    expect(component.find('.sub__text')).toHaveLength(2);
  });
});
