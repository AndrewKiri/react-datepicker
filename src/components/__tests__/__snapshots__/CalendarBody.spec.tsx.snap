// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CalendarBody /> prop: viewMode should ViewMode.DAY correctly 1`] = `
<div
  className="calendar__body"
>
  <DayView
    current={"2018-12-01T00:00:00.000Z"}
    locale="en-ca"
    onClick={[Function]}
    viewMode={2}
  />
</div>
`;

exports[`<CalendarBody /> should render correctly 1`] = `
<CalendarBody
  current={"2018-12-01T00:00:00.000Z"}
  locale="en-ca"
  onClick={[Function]}
  viewMode={2}
/>
`;
